### PR TITLE
Haneef/token swap

### PR DESCRIPTION
**Pull request by haneefghanim**
Created Token Swap page.

Pulled in `react-router` for routing. Using Binance API to pull in dynamic prices from token -> USDT. Still need to pull in real wallet info as well as token -> token exchange rates from CLPs. Right now we have only one CLP running for XLM, will need to add all the ones in the `helpers/tokens.ts` file.

### CODE REVIEW

File: gui/src/helpers/getInUsd.ts [hunk 1]
```diff
@@ -1,4 +1,4 @@
-// TODO replace with dynamic function by Haneef
+// TODO usage of this function with getTokenPriceInUsd in Store.ts
```
Comment (philipstanislaus) [inline gui/src/helpers/getInUsd.ts]: Shall I take care of this?

File: gui/src/organisms/CreateSwap.tsx [hunk 2]
```diff
@@ -0,0 +1,310 @@
...
+    selectedExchangeAmount: null,
+    selectedExchangePercentage: 100,
+    selectedExchangeToken: null,
+    selectedReceiveToken: null,
+  }
+
+  constructor(props:IProps) {
+    super(props)
+    props.store!.fetchPrices()
+    this.handleExchangePercentageClick = this.handleExchangePercentageClick.bind(this)
```
Comment (philipstanislaus) [inline gui/src/organisms/CreateSwap.tsx]: Please use arrow functions for better readability: 

```ts
public handleExchangePercentageClick = (selectedExchangePercentage:number) =>  {
    this.setState({
      selectedExchangePercentage,
    })
  }
```

File: gui/src/organisms/CreateSwap.tsx [hunk 3]
```diff
@@ -0,0 +1,310 @@
...
+      },
+    ]
+
+    const exchangeTokensToRender = [
+      { denom: 'BTC' },
+      { denom: 'ETH' },
+      { denom: 'LTC' },
+      { denom: 'XMR' },
+      { denom: 'DASH' },
+      // { denom: 'BCH' }, TODO uncomment when hard fork of BCH has been resolved
```
Comment (philipstanislaus) [inline gui/src/organisms/CreateSwap.tsx]: Well done ;)

File: gui/src/organisms/CreateSwap.tsx [hunk 4]
```diff
@@ -0,0 +1,310 @@
...
+            <Header>
+              Your Wallet
+            </Header>
+            {walletTokens.map((token) => (
+              <CoinWrapper key={token.denom}>
+                <Coin
+                  type={token.denom}
+                  selected={selectedExchangeToken === token.denom}
+                  style={{ marginRight: 10 }}
+                  onClick={this.handleExchangeTokenClick.bind(this, token.denom, Number(token.amount))}
```
Comment (philipstanislaus) [inline gui/src/organisms/CreateSwap.tsx]: Bind here is also unnecessary if you use arrow functions above.

File: gui/src/store/Store.ts [hunk 5]
```diff
@@ -119,14 +118,68 @@ const Pair = types.model({
...
+  fetchPrices: flow(function* fetchPrices() {
+    try {
+      const prices: [IPrice] = yield http.get(
+        `${env.REACT_APP_API_HOST}/swap/prices`,
+      )
+      self.prices = cast(prices.map(price => ({ symbol: price.symbol, price: Number(price.price) })))
+    } catch (error) {
+      // tslint:disable-next-line:no-console
+      console.error(`Failed to fetch prices`, error)
+    }
```
Comment (philipstanislaus) [inline gui/src/store/Store.ts]: Shall we regularly refresh these prices, e. g. every 5 seconds? Later we probably want to subscribe via Websockets – but for now polling is fine IMHO.

File: gui/src/store/Store.ts [hunk 6]
```diff
@@ -119,14 +118,68 @@ const Pair = types.model({
...
+  }),
+  getTokenPriceInUsd(amount: number, denom: string) {
```
Comment (philipstanislaus) [inline gui/src/store/Store.ts]: Let's call this `getTokenPriceInUsdt` to communicate that it's USDT and not USD that's fetched, okay?

File: gui/src/store/Store.ts [hunk 7]
```diff
@@ -119,14 +118,68 @@ const Pair = types.model({
...
+    const pricesData = self.prices
+
+    if (!pricesData) {
+      return 0
```
Comment (philipstanislaus) [inline gui/src/store/Store.ts]: Let's use `null` to indicate we have no data – what do you think?

File: gui/src/store/Store.ts [hunk 8]
```diff
@@ -119,14 +118,68 @@ const Pair = types.model({
...
+      if (price.symbol === 'BTCUSDT') {
+        BTCtoUSDT = price.price
+      }
+    }
+
+    if (BTCtoUSDT && denomToBTC) {
+      return amount * denomToBTC * BTCtoUSDT
+    }
+
+    return 0
```
Comment (philipstanislaus) [inline gui/src/store/Store.ts]: Here also: return `null`?

File: gui/src/store/Store.ts [hunk 9]
```diff
@@ -119,14 +118,68 @@ const Pair = types.model({
...
+  },
```
Comment (philipstanislaus) [inline gui/src/store/Store.ts]: `getTokenPriceInUsd` get's called very often – might want to cache responses in an object or memoize later – but don't worry fow now.

File: gui/src/helpers/getInUsd.ts (hunk 1 above)
Comment (haneefghanim) [reply gui/src/helpers/getInUsd.ts]: @philipstanislaus Yeah i figured that made the most sense

File: gui/src/organisms/CreateSwap.tsx (hunk 3 above)
Comment (haneefghanim) [reply gui/src/organisms/CreateSwap.tsx]: 😄

File: gui/src/organisms/CreateSwap.tsx (hunk 4 above)
Comment (haneefghanim) [reply gui/src/organisms/CreateSwap.tsx]: @philipstanislaus This is binding a different set of args to the `handleExchangeTokenClick` function based on the `walletTokens` that are being iterated on. Think I should change this to an arrow function for readability though?

File: gui/src/store/Store.ts (hunk 5 above)
Comment (haneefghanim) [reply gui/src/store/Store.ts]: @philipstanislaus Yep, good idea.

File: gui/src/store/Store.ts (hunk 7 above)
Comment (haneefghanim) [reply gui/src/store/Store.ts]: 👍

File: gui/src/store/Store.ts (hunk 8 above)
Comment (haneefghanim) [reply gui/src/store/Store.ts]: 👍

**Comment by Jessicawatsonmiller** [comment]
Still need to fix codeclimate issues @haneefghanim; but we can push these to separate tasks in a new sprint.

**Pull request merged**